### ISSUE TITLE
incompatible types: scala.Function1<A,C> cannot be converted to scala.Function1<A,B>

### ISSUE THREAD

**Issue by hohserg1** [scala/bug#11181]
```Scala
//Main
class A
class B
class C extends B

object Main extends App {
  def test(f:A=>B): Unit ={
    println(f(new A))
  }
  def t1= (_:A) =>new B
  def t2= (_:A) =>new C
 
  test(t1)//compile
  test(t2)//compile
}
 
 
//JavaTest
public class JavaTest {
    static public void main(String[] args){
        Main.test(Main.t1());//compile
        Main.test((A a)->new C());//compile
        Main.test(Main.t2());//not compile: incompatible types: scala.Function1<A,C> cannot be converted to scala.Function1<A,B>
 
    }
}
```
scalaVersion := "2.12.7"

code not compile if function A=>C taken from another function

**Comment by NthPortal** [comment]
I believe this is because Java does not have a concept of variance (except for arrays).

The return types of `t1` and `t2` are:

```scala
def t1: A => B
def t2: A => C
```

Because `Function1` is defined as `Function1[-T1, +R]`, `A => C` is a subtype of `A => B` (because `C` is a subtype of `B`). Thus, you can pass `t2` as an argument to `test`.

Java however, not having variance for generics, only sees `Function1<T1, R>`; consequently, it does not see `Function1<A, C>` as a subtype of `Function1<A, B>`. Because `test` takes a `Function1<A, B>`, you cannot pass it `t2` - a `Function1<A, C>` as an argument.

**tl;dr** Scala's type system is more flexible than Java's, so not everything that works in Scala works in Java.

**Comment by hohserg1** [comment]
if 
```scala
Function1[A,B]
```
would be translated to
```java
Function1<? super A, ? extends B>
```
it solved this problem

**Comment by hohserg1** [comment]
Or, if 
```scala
Function1[-T1,+T2]
```
would be translated to
```java
Function1<T1, T2, t1 super T1, t2 extends T2>
```

**Comment by SethTisue** [comment]
But Java has use-site variance, not declaration-site variance. And it isn't clear where such an opportunity for something to be "translated" would even come, since we don't control javac.

Perhaps there is something we could improve here on the Scala side — if you have a design in mind, a good place to discuss it and get further clarity would be https://contributors.scala-lang.org
